Makefile

```diff
@@ -4,29 +4,116 @@
 #
 #  SPDX-License-Identifier: LGPL-3.0-or-later
 #
-#  Authors: anita.shekar@ibm.com, sandy.kaur@ibm.com
+#  Authors: anita.shekar@ibm.com, sandy.kaur@ibm.com, guptsanj@us.ibm.com
 #
-
+SHELL := /bin/bash
+PROJDIR := $(dir $(realpath $(firstword $(MAKEFILE_LIST))))
 MODULE := control
 CONFIG ?= ceph-nvmeof.conf
+curr_dir := $(shell pwd)
+DOCKER_NO_CACHE :=
+CONT_NAME := ceph-nvmeof
+CONT_VERS := latest
+REMOTE_REPO ?= ""
+SPDK_VERSION := $(shell cd spdk;git -C . describe --tags --abbrev=0 | sed -r 's/-/\./g')
+DOCKER_VERSION := $(shell docker image ls | grep spdk | tr -s ' ' | cut -d ' ' -f2)
+PYVENV := $(PROJDIR)/venv
+CEPH_VERSION := 17.2.5
+
+# Utility Function to activate the Python Virtual Environment
+define callpyvenv =
+	(source ${PYVENV}/bin/activate; export PATH=$(PYVENV)/bin; $(1))
+endef
 
-setup: requirements.txt
-	pip3 install -r requirements.txt
+## setup: setup add requirements
+.PHONY: setup
+setup:
+	@echo "dir: $(curr_dir)/spdk"
+	cd $(curr_dir)/spdk && \
+	git submodule update --init --recursive
 
-grpc:
+## grpc: Compile grpc code
+.PHONY: grpc
+grpc: $(PYVENV)
 	@mkdir -p $(MODULE)/generated
-	@python3 -m grpc_tools.protoc \
+	$(call callpyvenv, python3 -m grpc_tools.protoc \
 			--proto_path=./proto \
 			--python_out=./$(MODULE)/generated \
 			--grpc_python_out=./$(MODULE)/generated \
-			./proto/*.proto
+			./proto/*.proto)
 	@sed -i 's/^import.*_pb2/from . \0/' ./$(MODULE)/generated/*.py
 
-run:
-	@python3 -m $(MODULE) -c $(CONFIG)
+## run: Run the gateway server
+.PHONY: run
+run: $(PYVENV)
+	@echo "Executing PEP8 on python files..."
+	$(call callpyvenv,python3 -m $(MODULE) -c $(CONFIG))
+
+## test: Run tests
+.PHONY: test
+test: $(PYVENV)
+	$(call callpyvenv,pytest)
+
+## spdk-image: Build spdk image if it does not exist
+.PHONY: spdk-image
+spdk-image:
+	ln -sf docker/.dockerignore.spdk .dockerignore
+	docker buildx build \
+	--network=host \
+	--build-arg spdk_version=$(SPDK_VERSION) \
+	--build-arg CEPH_VERSION=$(CEPH_VERSION) \
+	--build-arg spdk_branch=ceph-nvmeof \
+	--progress=plain \
+	${DOCKER_NO_CACHE} \
+	-t spdk:$(SPDK_VERSION) -f docker/Dockerfile.spdk .
+	@touch .spdk-image
+
+## spdk-rpms: Copy the rpms from spdk container in output directory
+.PHONY: spdk-rpms
+spdk_rpms:
+	docker run --rm -v $(curr_dir)/output:/output spdk:$(SPDK_VERSION) \
+	bash -c "cp -f /tmp/rpms/*.rpm /output/"
+
 
-test:
-	@pytest
+## gateway-image: Build the ceph-nvme gateway image. The spdk image needs to be built first.
+.PHONY: gateway-image
+gateway-image: spdk-image
+	ln -sf  docker/.dockerignore.gateway .dockerignore
```

Comment:
nit: This symlink thing looks hackish. It looks like you could use `docker/Dockerfile.gateway.dockerignore` instead. See e.g. this example: https://github.com/moby/moby/issues/12886#issuecomment-480575928

Reply:
done